### ISSUE TITLE
Silent Installer for Client Side/Credential Provider

### ISSUE THREAD

**Issue by Jonathan-Garber** [multiOTP/multiotp#27]
Quick feature request here.

We need a silent installer for the credential provider, as this would make mass deployment/updating way easier on a domain. Right now I roll our own but it would be way easier to have an existing one so I don't have to keep updating ours for the latest exe and dlls etc.

**Comment by multiOTP** [comment]
Hello Jonathan, and thanks a lot using our open source solution.

Your suggestion for the creation of a MSI silent installer (probably using WIX) is already in our backlog and will be done in the future. Furthermore, the multiOTP.exe part is already able to read in priority configuration options that are stored in a registry entry instead of the configuration file.

If you are in a hurry for the silent installer, or If you simply want to support us, please be free to have a look at the main page of the project and click on the donation PayPal to prioritize your specific request.

Best regards,

The **multiOTP** open source team

**Comment by Jonathan-Garber** [comment]
I created an installer using INNO that we can deploy with command line switches to configure it. We deploy it using GPO as a start up script.

The startup script ensures the credential provider is always installed and present on the target system. If on startup the credential provider isn't found, it re-installs it. If the credential provider is present, the installer skips.

> Furthermore, the multiOTP.exe part is already able to read in priority configuration options that are stored in a registry entry instead of the configuration file.

I am already using the registry entry for configuration options, however it doesn't work with just the registry entries. multiOTP.exe still requires the INI file. I ran a test without the INI file and it fails to logon with a "User Doesn't Exist" error.

When I get back into the system I can see a new INI file was created automatically by multiOTP.exe. So I am assuming that the registry entry doesn't include allthe required variables for multiOTP.exe to function properly. If I had a list of all the variables required to function, I could push them to the reg entry and eliminate the INI file completely.

Ideally, I would use silent install to push everything to the registry based on command line arguements. This eliminates the need for the INI.

As it sits right now, if you look at the multiOTP.ini it is bloated with what I am sure are just unused variables and not required client side. If I can get a list of the absolute required variables I can clean up the config and work on an installer that utilizes the reg key entirely.

**Comment by multiOTP** [comment]
Hi Jonathan,
Are you already using the multiOTPOptions registry entry which has been introduced in 5.1.0.3 ?
Regards,
Andre

**Comment by Jonathan-Garber** [comment]
> I am already using the registry entry for configuration options, however it doesn't work with just the registry entries. multiOTP.exe still requires the INI file. I ran a test without the INI file and it fails to logon with a "User Doesn't Exist" error.

I am using the multiOTPOptions registry key and seeing the behavior I listed above.

**Comment by multiOTP** [comment]
Check with the last 5.1.0.4 version here : [https://download.multiotp.net/credential-provider/](https://download.multiotp.net/credential-provider/)
Regards

**Comment by Jonathan-Garber** [comment]
5.1.0.4 Logon Failed -- I receive "User Doesn't Exist" error when I try to logon with OTP.

I am **not including** the "multiotp.ini" that was created for our environment in the installer since it is supposed to take the settings it needs from the registry keys.

Here are the reg keys my installer creates.

```
[HKEY_CLASSES_ROOT\CLSID\{FCEFDFAB-B0A1-4C4D-8B2B-4FF4E0A3D978}]
@="multiOTPCredentialProvider"
"multiOTPCacheEnabled"=dword:00000001
"multiOTPDisplaySmsLink"=dword:00000000
"multiOTPLoginTitle"="Secure Logon"
"multiOTPOptions"="server_secret=SECRET    server_cache_level=1    server_timeout=10    server_url=https://otp.secure.com:8113;http://otp.secure.com:8112"
"multiOTPPath"="C:\\My Company\\Secure Logon\\"
"multiOTPPrefixPass"=dword:00000000
"multiOTPRDPOnly"=dword:00000000
"multiOTPServers"="https://otp.secure.com:8113;http://otp.secure.com:8112"
"multiOTPServerTimeout"=dword:0000000a
"multiOTPSharedSecret"="SECRET"
"multiOTPTimeout"=dword:0000000a
"multiOTPUPNFormat"=dword:00000000

[HKEY_CLASSES_ROOT\CLSID\{FCEFDFAB-B0A1-4C4D-8B2B-4FF4E0A3D978}\InprocServer32]
@="multiOTPCredentialProvider.dll"
"ThreadingModel"="Apartment"
```

Here is the multiotp.ini file it generated when I tried but failed to logon.

```multiotp-database-format-v3
; If backend is set to something different than files,
; and backend_type_validated is set to 1,
; only the specific information needed for the backend
; is used from this config file.

encryption_hash=REMOVED ( It DID have the same hash key here as the multiotp.ini from our environment so I removed it for this post )
server_secret:=dGx9cn5qeFV7fFJ5RGR/cGd4 ( Not sure how it encodes the server secret key so not sure if this is correct for the key "SECRET" )
actual_version=5.1.0.3
admin_password_hash:=
anonymous_stat=1
anonymous_stat_last_update=0
anonymous_stat_random_id=
attributes_to_encrypt=
auto_resync=1
backend_encoding=UTF-8
backend_type=files
backend_type_validated=0
cache_data=0
cache_ldap_hash=1
case_sensitive_users=0
clear_otp_attribute=
console_authentication=0
debug=0
default_algorithm=totp
default_dialin_ip_mask=
default_user_group=
default_request_ldap_pwd=1
default_request_prefix_pin=1
demo_mode=0
developer_mode=0
display_log=0
domain_name=
email_admin_address=
encode_file_id=0
encryption_key_full_path=
failure_delayed_time=300
group_attribute=Filter-Id
hash_salt_full_path=
issuer=multiOTP
language=en
last_update=1519229230
ldap_expired_password_valid=1
ldap_account_suffix=
ldap_activated=0
ldap_base_dn=
ldap_bind_dn=
ldap_cache_folder=tempdir
ldap_cache_on=1
ldap_cn_identifier=sAMAccountName
ldap_default_algorithm=totp
ldap_domain_controllers=
last_failed_white_delay=60
ldap_group_attribute=memberOf
ldap_group_cn_identifier=cn
ldap_groups_dn=
ldap_hash_cache_time=604800
ldap_in_group=
ldap_language_attribute=preferredLanguage
ldap_network_timeout=10
ldap_port=389
ldap_recursive_cache_only=0
ldap_recursive_groups=1
ldap_server_password:=
ldap_server_type=1
ldap_ssl=0
ldap_synced_user_attribute=
ldap_time_limit=30
ldaptls_reqcert=auto
ldaptls_cipher_suite=auto
log=0
max_block_failures=6
max_delayed_failures=3
max_event_resync_window=10000
max_event_window=100
max_time_resync_window=90000
max_time_window=600
multiple_groups=0
ntp_server=pool.ntp.org
overwrite_request_ldap_pwd=1
radius_error_reply_message=1
radius_reply_attributor= += 
radius_reply_separator_hex=2c
radius_tag_prefix=
scratch_passwords_digits=6
scratch_passwords_amount=10
self_registration=1
server_cache_level=1
server_cache_lifetime=15552000
server_timeout=5
server_type=xml
server_url=
sms_api_id:=
sms_message_prefix=%s is your SMS-Code
sms_originator=multiOTP
sms_password:=
sms_provider=
sms_userkey:=
sms_digits=6
sms_timeout=180
smtp_auth=0
smtp_password:=
smtp_port=25
smtp_sender=
smtp_sender_name=
smtp_server=
smtp_ssl=0
smtp_username=
sql_server=
sql_username=
sql_password:=
sql_database=
sql_schema=
sql_config_table=multiotp_config
sql_cache_table=multiotp_cache
sql_devices_table=multiotp_devices
sql_groups_table=multiotp_groups
sql_log_table=multiotp_log
sql_tokens_table=multiotp_tokens
sql_users_table=multiotp_users
syslog_facility=7
syslog_level=5
syslog_port=514
syslog_server=
tel_default_country_code=
timezone=Europe/Zurich
token_serial_number_length=12
token_otp_list_of_length=6
verbose_log_prefix=
```

The installer works perfectly in our environment as long as I include the multiotp.ini for our environment in the installer. It just seems to either ignore the reg entries or the reg entries are missing some other vital piece the multiotp.exe or credential provider needs.

The goal is to not need the INI file at all so the installer can be universally used by anyone. They simply configure their install for their environment via command line arguments. This effectively makes a mass deployment installer that can be deployed in any network via batch files, start up scripts etc.

Any ideas?

**Comment by multiOTP** [comment]
To avoid virus false positive alert, multiOTP.exe is NO more packaged in one single file using Enigma, a php folder is now included in the multiOTP folder. A new installation file is available for multiOTPCredentialProvider (https://github.com/multiOTP/multiOTPCredentialProvider/releases/tag/5.1.0.5)
We have tested the installation by keeping only the registry entries and deleteing the multiotp.ini file.
The authentication is successful and the multiotp.ini is created if missing.
Thanks for your feedback,
Regards

**Comment by Jonathan-Garber** [comment]
I would revert the changes to Engima and leave it packaged all in one .exe as this was a cleaner install. No need for excessive folders. The false virus alerts were already resolved, there were only a handful of anti-virus tools showing a false positive and I reported the false positive to Webroot who have already fixed it for theirs.

I am still getting "User Doesn't Exist" errors and unable to logon even with 5.1.0.5 so I am expecting there is an error on my end somewhere. My registry keys look correct but I am suspecting the key **multiOTPOptions** is where I am having issues.

I think the INNO installer is putting incorrect spacing between the variable entries and they are not being read correctly. The documentation says to use "TAB" spacing between each variable entry. It looks like INNO installer is writing it with single spaces ( 4 spaces).

I am going to test again after adjusting the registry entry to confirm that is the issue.

**Comment by multiOTP** [comment]
Hello Jonathan,
Do not use multiOTPOptions, it's for **additional** options only, the following registry entries are enough:
* multiOTPCacheEnabled
* multiOTPServers
* multiOTPServerTimeout
* multiOTPSharedSecret

**Comment by Jonathan-Garber** [comment]
Ok something is still not working right here. I still get "User Doesn't Exist"
Here is my installation

![otperrors](https://user-images.githubusercontent.com/2494571/36554564-9aaaff32-17cd-11e8-94c0-7466e471c736.PNG)

I discovered a really weird behavior during testing.
I get "User Doesn't Exist" error when we don't put our multiotp.ini config file in.
if I log back in under safe mode and remove the default multiotp.ini it created because I didn't include ours. Then I put ours back in, I am able to logon and it works.

Now If I delete our multiotp.ini config and lock my machine. I can still get back in and see the default multiotp.ini was created again. It magically works now with the default multiotp.ini file.


TL:DR
I cannot logon "User Doesn't Exist"
I put our multiotp.ini in config folder and it works
I then delete our multiotp.ini config from the folder and I can still logon.
When I logon and look the default multiotp.ini has been recreated but the system still works and lets me logon with it now.

Any ideas here LOL

**Comment by Jonathan-Garber** [comment]
I have also tested by removing the users DB file in "Users" folder and removing the config file. The "User Doesn't Exist" error returns when I do this.

So it needs our multiotp.ini in order to properly logon the first time which creates "Users" folder then the users DB file. Then after that it doesn't care what multiotp.ini file is there because the user DB file exists. No more "User Doesn't Exist" errors.

So it seems it still isn't utilizing registry keys correctly. It shouldn't need the multiotp.ini file to logon and create the users DB it should take the keys needed from registry. At least that is what I am seeing from all this testing.

**Comment by Jonathan-Garber** [comment]
Ok so I can confirm the following.

If a user doesn't exist in **Users** folder then the **multiotp.ini** file  is required and the following variables must be there and be accurate.
```
server_url=SERVER URL
server_secret:=SECRET
```
Once you logon and the user is created in the **Users** folder, then **multiotp.ini** doesn't matter anymore.

I deleted my **username.db** file, added my server secret and server url to the **default multiotp.ini** and I was able to logon successfully. Then I deleted the **default multiotp.ini** file and was still able to logon because my **user.db** file now exists.

TL:DR
**Multiotp.ini** is required with your correct **secret key** and **server url** in order to handle the **absolute first logon of a new user** to that system via local or rdp. Once that user has a **username.db** file located in the **Users** folder. the **multiotp.ini** file no longer matters and that user can logon without error.

I suspect the **multiotp.exe** is not properly reading the **multiOTPServers** or **multiOTPSharedSecret** keys from the **registry**. So it looks to the **multiotp.ini** file which if default doesn't have the correct **secret** or **server_url** entry and if the user attempting to logon doesn't have a **username.db** file, the logon fails with **User Doesn't Exist**

**Comment by multiOTP** [comment]
Hello,
We were able to reproduce the problem on one of our machine, the next release will contain a new implementation. The registry entries will be read by the Credential Provider itselfm and information will be passed to multiotp executable.
Regards,

**Comment by Jonathan-Garber** [comment]
Great news. Thanks.
Look forward to the update.

**Comment by multiOTP** [comment]
Hello Jonathan,

Please have a look here for the files of the 5.1.0.6 version of the Credential Provider : https://download.multiotp.net/credential-provider/multiOTPCredentialProvider-files-only-5.1.0.6.zip
You will get:
- the last multiOTP.exe (in a single file using Enigma, as requested)
- the i386 edition of the multiOTPCredentialProvider.dll
- the x64 edition of the multiOTPCredentialProvider.dll

The following registry entries under **HKEY_CLASSES_ROOT\CLSID\{FCEFDFAB-B0A1-4C4D-8B2B-4FF4E0A3D978}** are used by the multiOTP Credential Provider:
- multiOTPCacheEnabled (used to add a parameter when calling multiOTP.exe)
- multiOTPServers (used to add a parameter when calling multiOTP.exe)
- multiOTPServerTimeout (used to add a parameter when calling multiOTP.exe)
- multiOTPSharedSecret (used to add a parameter when calling multiOTP.exe)
- multiOTPDisplaySmsLink
- multiOTPLoginTitle
- multiOTPPath
- multiOTPPrefixPass
- multiOTPRDPOnly
- multiOTPTimeout
- multiOTPUPNFormat

Keep us in touch to know if everything is now working as expected on your side.

Regards,

Andre

**Comment by Jonathan-Garber** [comment]
I have not had a chance to test this new install yet. I will try to get to it this Monday and get you some feedback. Look forward to testing it.

**Comment by multiOTP** [comment]
Hello Jonathan,
Thanks for the information, we are waiting your feedback. If your feedback is positive with the 5.1.0.7 version, we will push at the top of our roadmap the .msi creation (using Wix Toolset to create it).
Regards,
Andre

**Comment by Jonathan-Garber** [comment]
Doesn't work.

"Cannot Load Library php7.dll"

**Comment by multiOTP** [comment]
Hello Jonathan,
:-( Bad issue with the Enigma Virtual Box packer and x64 architecture, I have updated to version 8.10 of Enigma Virtual Box,, the new "files only" zip 5.1.0.8 is available on https://download.multiotp.net/credential-provider/
Hope it's the good one.
Regards,
Andre

**Comment by Jonathan-Garber** [comment]
I haven't been able to test this yet, I am trying to get to it now and will post feedback soon.

**Comment by Jonathan-Garber** [comment]
This worked as expected.
Thanks.

**Comment by multiOTP** [comment]
Cool. We will provide an "official" msi silent installer in one of the next releases.
Regards
